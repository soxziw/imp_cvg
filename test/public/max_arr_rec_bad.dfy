// verified
method MaxInductionHelper(a: array<int>, i: int, len: int)
  returns (r: int)
  requires len > 0
  requires 0 <= i < len
  ensures forall j :: i <= j < len ==> a[j] <= r
  ensures exists j :: i <= j < len && a[j] == r
{
  var result : int;
  var prev : int;

  if i == len - 1 {
    result := a[i];
  } else {
    result := MaxInductionHelper(a, i + 1, len);
    if a[i] > result {
      result := a[i];
    } else {}
  }
  return result;
}

// verified
method MaxInduction(a: array<int>, len: int)
  returns (r: int)
  requires len > 0
  ensures forall i :: 0 <= i < len ==> a[i] <= r
  // ensures exists i :: 0 <= i < len && a[i] == r
{
  var result: int;
  result := MaxInductionHelper(a, 0, len);
  return result;
}

// verified
method MaxIterHelper(a: array<int>, i: int, curr: int, len: int)
  returns (r: int)
  requires len >= 0
  requires 0 <= i <= len
  requires forall j :: 0 <= j < i ==> a[j] <= curr
  requires exists j :: 0 <= j < i && a[j] == curr
  ensures forall i :: 0 <= i < len ==> a[i] <= r
  ensures exists i :: 0 <= i < len && a[i] == r
{
  var result : int;
  var next : int;
  if i == len {
    result := curr;
  } else {
    if a[i] > curr {
      next := a[i];
    } else {
      next := curr;
    }
    result := MaxIterHelper(a, i + 1, next, len);
  }
  return result;
}

// verified
method MaxIter(a: array<int>, len: int)
  returns (r: int)
  requires len > 0
  ensures forall i :: 0 <= i < len ==> a[i] <= r
  ensures exists i :: 0 <= i < len && a[i] == r
{
  var result : int;
  result := MaxIterHelper(a, 1, a[0], len);
  return result;
}

// not verified
method Main() returns (r: int) {
  var a: array<int>;
  var i : int;
  var n : int;
  var h : int;
  var result1 : int;
  var result2 : int;

  n := *;
  a := *;
  assume n > 0;

  result1 := MaxInduction(a, n);
  result2 := MaxIter(a, n);
  print result1;
  print result2;
  assert result1 == result2;
  return 0;
}