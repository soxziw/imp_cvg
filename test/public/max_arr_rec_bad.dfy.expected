MaxInductionHelper: verified
MaxInduction: verified
MaxIterHelper: verified
MaxIter: verified
Main: not verified
